##  Random forest 
### Random Forest Classification on Glass and Diabetis Datasets.
The goal of the project is to 
1.Find the type of glass whether it is "1" or "2" or "3" or "5" or "6" or "7" with given features like Ri,Na,Mg,AL,Si(types of chemicals used to make glass) and find the best parameters with grid search and random search cross validation to improve the accuracy.
2.Find the a patient whether diabetic class"1" or not class"0" using 

In [1]:
# installing required packages
!pip install scikit-learn

# Import a bunch of libraries.
import pandas as pd
import time
import numpy as np
import warnings
warnings.filterwarnings("ignore")
import sklearn
from sklearn.metrics import roc_curve, auc
from sklearn.pipeline import Pipeline
from sklearn.datasets import fetch_mldata
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report

### Data Extraction

Data extracted from Kaggle.Datset contains 214 records  with 10 features with float data type. Type is target variable. 

In [70]:
path = '/Users/mummadi/Downloads/glass.csv'
glass = pd.read_csv(path)
glass.head()

RI     Na    Mg    Al     Si     K    Ca   Ba   Fe  Type
0  1.52101  13.64  4.49  1.10  71.78  0.06  8.75  0.0  0.0     1
1  1.51761  13.89  3.60  1.36  72.73  0.48  7.83  0.0  0.0     1
2  1.51618  13.53  3.55  1.54  72.99  0.39  7.78  0.0  0.0     1
3  1.51766  13.21  3.69  1.29  72.61  0.57  8.22  0.0  0.0     1
4  1.51742  13.27  3.62  1.24  73.08  0.55  8.07  0.0  0.0     1

In [71]:
glass.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 214 entries, 0 to 213
Data columns (total 10 columns):
RI      214 non-null float64
Na      214 non-null float64
Mg      214 non-null float64
Al      214 non-null float64
Si      214 non-null float64
K       214 non-null float64
Ca      214 non-null float64
Ba      214 non-null float64
Fe      214 non-null float64
Type    214 non-null int64
dtypes: float64(9), int64(1)
memory usage: 16.8 KB


In [72]:
#  seperating target variable for  analysis
X = glass.drop('Type', axis=1)
y = glass['Type']

In [73]:
# split the data usin train, test, split method from sklearn.
from sklearn.model_selection import train_test_split
# implementing train-test-split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=66)

In [82]:
#imporing the randomforest classifier from sklearn ensemble.
from sklearn import model_selection
from sklearn.ensemble import RandomForestClassifier
# random forest model creation
rfc = RandomForestClassifier()
# training the model
rfc.fit(X_train,y_train)
# predictions
rfc_predict = rfc.predict(X_test)
metrics.accuracy_score(y_test,rfc_predict)

0.676056338028169

In [75]:
rfc_predict

array([1, 1, 2, 2, 5, 2, 2, 1, 1, 2, 1, 2, 1, 1, 7, 2, 2, 3, 2, 1, 7, 2,
       1, 3, 7, 1, 1, 2, 1, 2, 5, 5, 1, 1, 1, 2, 1, 1, 7, 2, 7, 1, 7, 2,
       1, 2, 5, 1, 1, 1, 2, 1, 1, 7, 1, 1, 1, 1, 1, 1, 1, 6, 2, 1, 1, 6,
       2, 1, 1, 2, 2])

In [81]:
# importing the classification report and confusion matrix from sklearn metrics
from sklearn.model_selection import cross_val_score
from sklearn.metrics import classification_report, confusion_matrix

# printing the confusion matrix statistics and classification report
print(confusion_matrix(y_test, rfc_predict))
print(classification_report(y_test, rfc_predict))

[[20  4  0  0  0  0]
 [ 4 17  0  1  0  1]
 [ 6  0  2  0  0  0]
 [ 0  1  0  3  0  0]
 [ 0  2  0  0  2  0]
 [ 0  1  0  0  0  7]]
              precision    recall  f1-score   support

           1       0.67      0.83      0.74        24
           2       0.68      0.74      0.71        23
           3       1.00      0.25      0.40         8
           5       0.75      0.75      0.75         4
           6       1.00      0.50      0.67         4
           7       0.88      0.88      0.88         8

   micro avg       0.72      0.72      0.72        71
   macro avg       0.83      0.66      0.69        71
weighted avg       0.76      0.72      0.70        71



In [79]:
# importing the random search cv for from model_selection for best parameter slection.
from sklearn.model_selection import RandomizedSearchCV
# number of trees in random forest
n_estimators = [int(x) for x in np.linspace(start = 200, stop = 2000, num = 10)]

# number of features at every split
max_features = ['auto', 'sqrt','log2']

# max depth
max_depth = [int(x) for x in np.linspace(100, 500, num = 10)]

# create random grid
random_grid = {
 'n_estimators': n_estimators,
 'max_features': max_features,
 'max_depth': max_depth
 }
# Random search of parameters
rfc_random = RandomizedSearchCV(estimator = rfc, param_distributions = random_grid, n_iter = 100, cv = 3, verbose=2, random_state=42, n_jobs = -1)
# Fit the model
rfc_random.fit(X_train, y_train)
# print results
print(rfc_random.best_params_)

Fitting 3 folds for each of 100 candidates, totalling 300 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=-1)]: Done  33 tasks      | elapsed:   15.0s
[Parallel(n_jobs=-1)]: Done 154 tasks      | elapsed:   57.1s


{'n_estimators': 200, 'max_features': 'auto', 'max_depth': 188}


[Parallel(n_jobs=-1)]: Done 300 out of 300 | elapsed:  1.8min finished
/Users/mummadi/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_search.py:841: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


In [80]:
# By giving best parameter checking the accuracy of prediction
rfc = RandomForestClassifier(n_estimators=200, max_depth=188, max_features='auto')
rfc.fit(X_train,y_train)
rfc_predict = rfc.predict(X_test)
metrics.accuracy_score(y_test,rfc_predict)

0.7183098591549296

In [27]:
print(confusion_matrix(y_test, rfc_predict))
print(classification_report(y_test, rfc_predict))

[[22  2  0  0  0  0]
 [ 3 17  0  2  0  1]
 [ 6  0  2  0  0  0]
 [ 0  1  0  3  0  0]
 [ 0  2  0  0  2  0]
 [ 0  1  0  0  0  7]]
              precision    recall  f1-score   support

           1       0.71      0.92      0.80        24
           2       0.74      0.74      0.74        23
           3       1.00      0.25      0.40         8
           5       0.60      0.75      0.67         4
           6       1.00      0.50      0.67         4
           7       0.88      0.88      0.88         8

   micro avg       0.75      0.75      0.75        71
   macro avg       0.82      0.67      0.69        71
weighted avg       0.78      0.75      0.73        71



With RandomsearchCv the best parameter got selected. It improved the accuracy from 67% to 71% still it is not very accurate.

## Diabetes Dataset
### Data Extraction

Data extracted from kaggle.Dataset contains 9 variables with 768 observations.Outcome is the target variable.

In [83]:
path = '/Users/mummadi/Downloads/diabetes.csv'
diabetis = pd.read_csv(path)
diabetis.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [84]:
diabetis.info()
# dataset looks clean

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
Pregnancies                 768 non-null int64
Glucose                     768 non-null int64
BloodPressure               768 non-null int64
SkinThickness               768 non-null int64
Insulin                     768 non-null int64
BMI                         768 non-null float64
DiabetesPedigreeFunction    768 non-null float64
Age                         768 non-null int64
Outcome                     768 non-null int64
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [85]:
#  seperating target variable for  analysis
X = diabetis.drop('Outcome', axis=1)
y = diabetis['Outcome']

In [86]:
# splitting the data with 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=66)

In [87]:
# random forest model creation
rfc = RandomForestClassifier()
rfc.fit(X_train,y_train)
# predictions
rfc_predict = rfc.predict(X_test)

In [88]:
rfc_cv_score = cross_val_score(rfc, X, y, cv=10, scoring='roc_auc')

In [89]:
from sklearn import metrics
scores = metrics.accuracy_score(y_test,rfc_predict)
print(scores)

0.7559055118110236


In [90]:
print("=== Confusion Matrix ===")
print(confusion_matrix(y_test, rfc_predict))
print('\n')
print("=== Classification Report ===")
print(classification_report(y_test, rfc_predict))
print('\n')
print("=== All AUC Scores ===")
print(rfc_cv_score)
print('\n')
print("=== Mean AUC Score ===")
print("Mean AUC Score - Random Forest: ", rfc_cv_score.mean())

=== Confusion Matrix ===
[[151  25]
 [ 37  41]]


=== Classification Report ===
              precision    recall  f1-score   support

           0       0.80      0.86      0.83       176
           1       0.62      0.53      0.57        78

   micro avg       0.76      0.76      0.76       254
   macro avg       0.71      0.69      0.70       254
weighted avg       0.75      0.76      0.75       254



=== All AUC Scores ===
[0.79814815 0.82333333 0.79185185 0.64518519 0.76037037 0.81777778
 0.84111111 0.82888889 0.70807692 0.80346154]


=== Mean AUC Score ===
Mean AUC Score - Random Forest:  0.7818205128205129


From the above, out of 176 non diabetes model predicted only 151 correct and outof 78 diabetes cases it is predicted only 41 correct. Model is not predicting diabetes class properly which is very important.

In [91]:
from sklearn.model_selection import GridSearchCV
# Create the parameter grid based on the results of random search 
param_grid = {
    'bootstrap': [True],
    'max_depth': [80, 90, 100, 110],
    'max_features': [2, 3],
    'min_samples_leaf': [3, 4, 5],
    'min_samples_split': [8, 10, 12],
    'n_estimators': [100, 200, 300, 1000]
}

# Instantiate the grid search model
grid_search = GridSearchCV(estimator = rfc, param_grid = param_grid, 
                          cv = 3, n_jobs = -1, verbose = 2)

In [92]:
# Fit the grid search to the data
grid_search.fit(X_train,y_train)

Fitting 3 folds for each of 288 candidates, totalling 864 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=-1)]: Done  33 tasks      | elapsed:    5.3s
[Parallel(n_jobs=-1)]: Done 154 tasks      | elapsed:   22.8s
[Parallel(n_jobs=-1)]: Done 357 tasks      | elapsed:   51.8s
[Parallel(n_jobs=-1)]: Done 640 tasks      | elapsed:  1.5min
[Parallel(n_jobs=-1)]: Done 864 out of 864 | elapsed:  2.1min finished


GridSearchCV(cv=3, error_score='raise-deprecating',
       estimator=RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=None,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False),
       fit_params=None, iid='warn', n_jobs=-1,
       param_grid={'bootstrap': [True], 'max_depth': [80, 90, 100, 110], 'max_features': [2, 3], 'min_samples_leaf': [3, 4, 5], 'min_samples_split': [8, 10, 12], 'n_estimators': [100, 200, 300, 1000]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=2)

In [93]:
grid_search.best_params_

{'bootstrap': True,
 'max_depth': 80,
 'max_features': 3,
 'min_samples_leaf': 3,
 'min_samples_split': 8,
 'n_estimators': 200}

In [94]:
rfc1=RandomForestClassifier(random_state=42, max_features=3, n_estimators= 200, max_depth=80, min_samples_leaf= 3,min_samples_split = 8)

In [95]:
rfc1.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=80, max_features=3, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=3, min_samples_split=8,
            min_weight_fraction_leaf=0.0, n_estimators=200, n_jobs=None,
            oob_score=False, random_state=42, verbose=0, warm_start=False)

In [96]:
pred=rfc1.predict(X_test)

In [97]:
pred

array([0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1,
       0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1,
       1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1,
       0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0,
       0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0,
       1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0,
       0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1])

In [100]:
print("=== Confusion Matrix ===")
print(confusion_matrix(y_test,pred))
print('\n')
print("=== Classification Report ===")
print(classification_report(y_test,pred))
print('\n')
print("=== All AUC Scores ===")
print(rfc_cv_score)
print('\n')
print("=== Mean AUC Score ===")
print("Mean AUC Score - Random Forest: ", rfc_cv_score.mean())
print("Accuracy for Random Forest on CV data: ",metrics.accuracy_score(y_test,pred))

=== Confusion Matrix ===
[[149  27]
 [ 32  46]]


=== Classification Report ===
              precision    recall  f1-score   support

           0       0.82      0.85      0.83       176
           1       0.63      0.59      0.61        78

   micro avg       0.77      0.77      0.77       254
   macro avg       0.73      0.72      0.72       254
weighted avg       0.76      0.77      0.77       254



=== All AUC Scores ===
[0.79814815 0.82333333 0.79185185 0.64518519 0.76037037 0.81777778
 0.84111111 0.82888889 0.70807692 0.80346154]


=== Mean AUC Score ===
Mean AUC Score - Random Forest:  0.7818205128205129
Accuracy for Random Forest on CV data:  0.7677165354330708


Fom above report, after gridserach for best parameter selection, the accuracy of the model not increased much. here the important class here diabetes class. it is giving only 76% accuracy. 

## Summary

### Glass dataset:

By using Random forest classifier, got just 67% accuracy. Miodel is not predicting accurately.with random grid selection model accuracy got increased to 71%. still it is not predicting greatly. for this dataset for multi class classification we can XGboost to get the better accuracy.

### Diabetes Datset:

By using Random forest Classifier, got just 75% accuracy. with gridsearch cv also also it is not improved much. For this Random forest is npt preferrable because it is predicting the positive (diabetes) class properly.which is very imporatnt.
